src/metabase/api/database.clj

```diff
@@ -404,54 +404,106 @@
-              [table-name "Table"])
-            (for [{:keys [table_name base_type semantic_type name]} (take take-fields fields)]
-              [name (str table_name
-                         " "
-                         base_type
-                         (when semantic_type
-                           (str " " semantic_type)))]))))
-
-(defn- autocomplete-suggestions
-  "match-string is a string that will be used with ilike. The it will be lowercased by autocomplete-{tables,fields}. "
-  [db-id match-string]
-  (let [limit  50
-        tables (filter mi/can-read? (autocomplete-tables db-id match-string limit))
-        fields (readable-fields-only (autocomplete-fields db-id match-string limit))]
-    (autocomplete-results tables fields limit)))
+(defn- autocomplete-card-columns
+  "Returns a list of card column metadata given the `search-string` and a set of card ids.
+   The search-string is matched against column names in the result_metadata of the cards.
+  Each result contains all the metadata for the column, and the name of the card."
+  [db-id search-string limit match-type card-ids]
+  (when (seq card-ids)
+    (->> (db/select [Card :name :result_metadata :collection_id]
+           :%lower.result_metadata         [:like (match-search-string :substring search-string)]
+           :database_id                    db-id
+           {:where [:in :id card-ids]
+            :limit limit})
+         (filter mi/can-read?)
+         (mapcat (fn [{:keys [result_metadata name]}]
+                   (map (fn [column-metadata]
+                          (merge column-metadata
+                                 {:card_name name}))
+                        result_metadata)))
+         (filter (fn [column-metadata]
+                   (re-find (re-pattern (case match-type
+                                          :prefix    (str "(?i)^" search-string)
+                                          :substring (str "(?i)" search-string))) (:name column-metadata))))
+         (sort-by :name))))
+
+(defmulti format-autocomplete-result
+  "Format an autocomplete result for the client."
+  {:arglists '([result-data])}
+  (fn [[result-type _]]
+    result-type))
+
+(defmethod format-autocomplete-result :table
+  [[_ {:keys [name]}]]
+  [name "Table"])
+
+(defmethod format-autocomplete-result :field
+  [[_ {:keys [table_name base_type semantic_type name]}]]
+  [name (str table_name
+             " "
+             base_type
+             (when semantic_type
+               (str " " semantic_type)))])
+
+(defmethod format-autocomplete-result :card-column
+  [[_ {:keys [card_name base_type semantic_type name]}]]
+  [name (str card_name
+             " "
+             base_type
+             (when semantic_type
+               (str " " semantic_type)))])
+
+(defn- autocomplete-results
+  "Returns ordered and formatted autocomplete results for the client to display."
+  [tables fields card-columns limit]
+  (->> (concat (map #(vector :card-column %) card-columns)
```

Comment:
1. how are we currently sorting the result between card/table/field, do we have priority orders for each of them?
2. I think you removed the notion of limit per type. Can you share why?

Reply:
Good questions. I should have documented this.
1. Tables > fields was the historical ranking. And Luiz answered in the notion doc that card-columns > tables or fields. Since we know the card is used in the query, we have higher confidence the user is going to use its columns vs the other types of results.
2. I made the decision to have a limit per type because it didn't make logical sense. My reasoning follows. Results should be ordered by the probability the user wants to use the result, with highest probability at the top, and lowest at the bottom. That is the order that minimises the expected amount of scrolling to find the desired result. The ordering of results within each type (card/table/field) is alphabetical, so in theory, the probability of the top card-column result is about the same as the bottom card-column result. And if we have decided that users are more likely to want a card-column than a table, then ALL the potential card-columns have a higher probability than the tables. So ALL the potential card-columns should be listed before tables, etc. So having a limit per type is suboptimal for the user when using alphabetical ordering. Also, it makes the code harder to understand and more likely to have bugs.